### PR TITLE
STORM-1910  One topology cannot use hdfs spout to read from two locations

### PR DESCRIPTION
**Pull request by roshannaik**
Changing the way the spout is configured. Using member functions as the primary mode of config... still have to supporting the older mode  (via conf object ) for backward compatibility.
Updated test code.

### CODE REVIEW

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 1]
```diff
@@ -291,10 +361,11 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
     this.tupleCounter = 0;
 
     // Hdfs related settings
-    if( conf.containsKey(Configs.HDFS_URI)) {
+    if(this.hdfsUri==null && conf.containsKey(Configs.HDFS_URI)) {
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: can you space after if

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 2]
```diff
@@ -291,10 +361,11 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
...
       this.hdfsUri = conf.get(Configs.HDFS_URI).toString();
-    } else {
-      throw new RuntimeException(Configs.HDFS_URI + " setting is required");
+    }
+    if(this.hdfsUri==null) {
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: space after if

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 3]
```diff
@@ -322,33 +393,41 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
       }
 
     // Reader type config
-    if( conf.containsKey(Configs.READER_TYPE) ) {
+    if( readerType==null && conf.containsKey(Configs.READER_TYPE) ) {
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: space after if

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 4]
```diff
@@ -322,33 +393,41 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
...
-      checkValidReader(readerType);
     }
+    checkValidReader(readerType);
 
     // -- source dir config
-    if ( !conf.containsKey(Configs.SOURCE_DIR) ) {
+    if ( sourceDir==null && conf.containsKey(Configs.SOURCE_DIR) ) {
+      sourceDir = conf.get(Configs.SOURCE_DIR).toString();
+    }
+    if( sourceDir==null ) {
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: space after if

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 5]
```diff
@@ -357,9 +436,15 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
     }
 
     // -- lock dir config
-    String lockDir = !conf.containsKey(Configs.LOCK_DIR) ? getDefaultLockDir(sourceDirPath) : conf.get(Configs.LOCK_DIR).toString() ;
+    if( lockDir==null && conf.containsKey(Configs.LOCK_DIR) ) {
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: remove the new line after the if and space after if

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 6]
```diff
@@ -357,9 +436,15 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
...
+      lockDir = conf.get(Configs.LOCK_DIR).toString();
+    }
+    if(lockDir==null) {
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: space after if

File: external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java [hunk 7]
```diff
@@ -357,9 +436,15 @@ public void open(Map conf, TopologyContext context, SpoutOutputCollector collect
...
+      lockDir = getDefaultLockDir(sourceDirPath);
+    }
     this.lockDirPath = new Path(lockDir);
-    validateOrMakeDir(hdfs,lockDirPath,"locks");
+    validateOrMakeDir(hdfs,lockDirPath, "locks");
```
Comment (harshach) [inline external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java]: space after hdfs,

**Comment by harshach** [comment]
minor nit picks. overall looks good to me +1 after addressing the  comments

**Comment by ptgoetz** [comment]
+1

**Pull request merged**